### ISSUE TITLE
Issues with Send & Receive Tasks, Boundary events and Timer Cycles.

### ISSUE THREAD

**Issue by MaheshkumarSundaram** [ralphhanna/bpmn-server#58]
**_Using 1.3.8_**

I'm using the below example to test Send & Receive Tasks.

![image](https://user-images.githubusercontent.com/55505611/189081118-896ab25c-0177-4f69-bbba-03b17452c894.png)

A) Interrupting boundary event - Timer Duration
B) Non-Interrupting boundary event - Timer Cycle

<details>
<summary>BPMN File</summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<bpmn2:definitions xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:bpmn2="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" xmlns:camunda="http://camunda.org/schema/1.0/bpmn" id="sample-diagram" targetNamespace="http://bpmn.io/schema/bpmn" xsi:schemaLocation="http://www.omg.org/spec/BPMN/20100524/MODEL BPMN20.xsd">
  <bpmn2:message id="Message_1c5ag87" name=" TestReceiveMsg" />
  <bpmn2:collaboration id="Collaboration_1scj0fl">
    <bpmn2:participant id="Participant_0murubu" processRef="Process_1" />
  </bpmn2:collaboration>
  <bpmn2:process id="Process_1" isExecutable="false">
    <bpmn2:startEvent id="StartEvent_1" name="Start">
      <bpmn2:outgoing>Flow_1vbusiz</bpmn2:outgoing>
    </bpmn2:startEvent>
    <bpmn2:userTask id="Activity_1wzqauq" name="Get Data">
      <bpmn2:extensionElements>
        <camunda:inputOutput>
          <camunda:outputParameter name="Name" />
          <camunda:outputParameter name="Age" />
        </camunda:inputOutput>
      </bpmn2:extensionElements>
      <bpmn2:incoming>Flow_1vbusiz</bpmn2:incoming>
      <bpmn2:outgoing>Flow_049jnj8</bpmn2:outgoing>
    </bpmn2:userTask>
    <bpmn2:receiveTask id="Activity_0lijns0" name="Receive" messageRef="Message_1c5ag87">
      <bpmn2:incoming>Flow_049jnj8</bpmn2:incoming>
      <bpmn2:outgoing>Flow_1k69s2j</bpmn2:outgoing>
    </bpmn2:receiveTask>
    <bpmn2:endEvent id="Event_0rhzt3f" name="End">
      <bpmn2:incoming>Flow_0nsp97x</bpmn2:incoming>
    </bpmn2:endEvent>
    <bpmn2:serviceTask id="Activity_1gca7eg" name="Dummy" camunda:delegateExpression="dummyService">
      <bpmn2:incoming>Flow_1k69s2j</bpmn2:incoming>
      <bpmn2:outgoing>Flow_0nsp97x</bpmn2:outgoing>
    </bpmn2:serviceTask>
    <bpmn2:boundaryEvent id="Event_0khrsn8" attachedToRef="Activity_0lijns0">
      <bpmn2:outgoing>Flow_00j2800</bpmn2:outgoing>
      <bpmn2:timerEventDefinition id="TimerEventDefinition_0zj43lm">
        <bpmn2:timeDuration xsi:type="bpmn2:tFormalExpression">PT1M</bpmn2:timeDuration>
      </bpmn2:timerEventDefinition>
    </bpmn2:boundaryEvent>
    <bpmn2:boundaryEvent id="Event_1jmzrm3" cancelActivity="false" attachedToRef="Activity_0lijns0">
      <bpmn2:outgoing>Flow_0qrn6iu</bpmn2:outgoing>
      <bpmn2:timerEventDefinition id="TimerEventDefinition_1gkftm0">
        <bpmn2:timeCycle xsi:type="bpmn2:tFormalExpression">R5/PT15S</bpmn2:timeCycle>
      </bpmn2:timerEventDefinition>
    </bpmn2:boundaryEvent>
    <bpmn2:sequenceFlow id="Flow_00j2800" sourceRef="Event_0khrsn8" targetRef="Activity_1p4flr2" />
    <bpmn2:sequenceFlow id="Flow_0qrn6iu" sourceRef="Event_1jmzrm3" targetRef="Activity_0k9j60h" />
    <bpmn2:sequenceFlow id="Flow_0nsp97x" sourceRef="Activity_1gca7eg" targetRef="Event_0rhzt3f" />
    <bpmn2:sequenceFlow id="Flow_1k69s2j" sourceRef="Activity_0lijns0" targetRef="Activity_1gca7eg" />
    <bpmn2:sequenceFlow id="Flow_049jnj8" sourceRef="Activity_1wzqauq" targetRef="Activity_0lijns0" />
    <bpmn2:sequenceFlow id="Flow_1vbusiz" sourceRef="StartEvent_1" targetRef="Activity_1wzqauq" />
    <bpmn2:serviceTask id="Activity_1p4flr2" name="Expired" camunda:delegateExpression="expiredService">
      <bpmn2:incoming>Flow_00j2800</bpmn2:incoming>
    </bpmn2:serviceTask>
    <bpmn2:sendTask id="Activity_0k9j60h" name="Notify" camunda:delegateExpression="notifyService">
      <bpmn2:incoming>Flow_0qrn6iu</bpmn2:incoming>
    </bpmn2:sendTask>
    <bpmn2:textAnnotation id="TextAnnotation_1ufv7fx">
      <bpmn2:text>After 1 min (PT1M)</bpmn2:text>
    </bpmn2:textAnnotation>
    <bpmn2:association id="Association_18y900z" sourceRef="Event_0khrsn8" targetRef="TextAnnotation_1ufv7fx" />
    <bpmn2:textAnnotation id="TextAnnotation_0aa6r3m">
      <bpmn2:text>Every 15 sec upto 5 times (R5/PT15S)</bpmn2:text>
    </bpmn2:textAnnotation>
    <bpmn2:association id="Association_1gbeptk" sourceRef="Event_1jmzrm3" targetRef="TextAnnotation_0aa6r3m" />
  </bpmn2:process>
  <bpmndi:BPMNDiagram id="BPMNDiagram_1">
    <bpmndi:BPMNPlane id="BPMNPlane_1" bpmnElement="Collaboration_1scj0fl">
      <bpmndi:BPMNShape id="Participant_0murubu_di" bpmnElement="Participant_0murubu" isHorizontal="true">
        <dc:Bounds x="220" y="80" width="1010" height="310" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="TextAnnotation_1ufv7fx_di" bpmnElement="TextAnnotation_1ufv7fx">
        <dc:Bounds x="530" y="257" width="100" height="41" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="TextAnnotation_0aa6r3m_di" bpmnElement="TextAnnotation_0aa6r3m">
        <dc:Bounds x="780" y="250" width="100" height="55" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge id="Flow_1vbusiz_di" bpmnElement="Flow_1vbusiz">
        <di:waypoint x="308" y="200" />
        <di:waypoint x="440" y="200" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_049jnj8_di" bpmnElement="Flow_049jnj8">
        <di:waypoint x="540" y="200" />
        <di:waypoint x="640" y="200" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_1k69s2j_di" bpmnElement="Flow_1k69s2j">
        <di:waypoint x="740" y="200" />
        <di:waypoint x="910" y="200" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_0nsp97x_di" bpmnElement="Flow_0nsp97x">
        <di:waypoint x="1010" y="200" />
        <di:waypoint x="1172" y="200" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_0qrn6iu_di" bpmnElement="Flow_0qrn6iu">
        <di:waypoint x="720" y="258" />
        <di:waypoint x="720" y="320" />
        <di:waypoint x="1040" y="320" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_00j2800_di" bpmnElement="Flow_00j2800">
        <di:waypoint x="670" y="258" />
        <di:waypoint x="670" y="320" />
        <di:waypoint x="480" y="320" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNShape id="_BPMNShape_StartEvent_2" bpmnElement="StartEvent_1">
        <dc:Bounds x="272" y="182" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="278" y="225" width="25" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_0ehe12m_di" bpmnElement="Activity_1wzqauq">
        <dc:Bounds x="440" y="160" width="100" height="80" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_0z8sjt1_di" bpmnElement="Activity_0lijns0">
        <dc:Bounds x="640" y="160" width="100" height="80" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Event_0rhzt3f_di" bpmnElement="Event_0rhzt3f">
        <dc:Bounds x="1172" y="182" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="1180" y="225" width="20" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_0jkmj4o_di" bpmnElement="Activity_1gca7eg">
        <dc:Bounds x="910" y="160" width="100" height="80" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_0v3m7y1_di" bpmnElement="Activity_1p4flr2">
        <dc:Bounds x="380" y="280" width="100" height="80" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_0qaisqb_di" bpmnElement="Activity_0k9j60h">
        <dc:Bounds x="1040" y="280" width="100" height="80" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge id="Association_18y900z_di" bpmnElement="Association_18y900z">
        <di:waypoint x="653" y="246" />
        <di:waypoint x="622" y="257" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Association_1gbeptk_di" bpmnElement="Association_1gbeptk">
        <di:waypoint x="738" y="244" />
        <di:waypoint x="780" y="254" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNShape id="Event_0rm859j_di" bpmnElement="Event_1jmzrm3">
        <dc:Bounds x="702" y="222" width="36" height="36" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Event_0a2rfq3_di" bpmnElement="Event_0khrsn8">
        <dc:Bounds x="652" y="222" width="36" height="36" />
      </bpmndi:BPMNShape>
    </bpmndi:BPMNPlane>
  </bpmndi:BPMNDiagram>
</bpmn2:definitions>

```
</details>

### Instance Execution:
![image](https://user-images.githubusercontent.com/55505611/189081495-00b46d26-693c-4fd8-847e-d225832eb003.png)

### Issues:

- Interrupting boundary events are not working as expected. After an Interrupting boundary event (here A) gets executed, I expect the existing task (here `Receive` task) to be cancelled. Isn't that the valid scenario as it's an interrupting one?
- I believe the send task (here `Notify` task) should behave like a service task. But it's service never gets called although the implementation is defined in `appDelegate`.
- Timer cycle (here B) events are not working.

Could you kindly look into this?
Thanks!

**Comment by ralphhanna** [comment]
Hi,
Unfortunately, the package that I use for timer **iso8601-duration** does not support repeat timers. Would you like to help and produce a TS Class to parse ISO8601 using Regex?
Thanks

**Comment by MaheshkumarSundaram** [comment]
Hi @ralphhanna,

Yeah. I would be glad to help with the timer cycle. I'm going on holidays for the next whole week. After I came back, I'll surely help with the timer.

Would you be fixing these two?
>> Interrupting boundary events are not working as expected. After an Interrupting boundary event (here A) gets executed, I expect the existing task (here Receive task) to be cancelled. Isn't that the valid scenario as it's an interrupting one?

>> I believe the send task (here Notify task) should behave like a service task. But it's service never gets called although the implementation is defined in appDelegate.

Thanks!

**Comment by ralphhanna** [comment]
Regarding: 
- Interrupting boundary events are not working as expected. After an Interrupting boundary event (here A) gets executed, I expect the existing task (here Receive task) to be cancelled. Isn't that the valid scenario as it's an interrupting one?
**You are correct, is fixed in next release**
- I believe the send task (here Notify task) should behave like a service task. But it's service never gets called although the implementation is defined in appDelegate.
 **You are correct, is fixed in next release**

Next release coming tonight.
Thanks

**Comment by MaheshkumarSundaram** [comment]
Great! Thank you!

**Comment by ralphhanna** [comment]
A bit of debate goes here:

![image](https://user-images.githubusercontent.com/11893416/189238653-b16a1653-7214-4cb9-864e-aa5ec0efbd6e.png)

When the first timer **Expired Timer** fires after specified time (1 minute) it will:
- End the Parent Activity **Receive**
- Which in turn will Cancel the **Notify Timer**
- Will invoke **Expired** ServiceTask
- Now the debate is: Should the **Receive** activity continue to execute **Dummy** Task 
- Or Should it Abort? in this case, the **Expired** Task need to take over processing

I am leaning to the first option, since you have control to make the flow to **Dummy** conditional, by setting a variable the Task Expired.

Just food for thoughts

**Comment by ralphhanna** [comment]
But then, this is a more clean option:
![image](https://user-images.githubusercontent.com/11893416/189241641-795e6262-4b5e-4862-91fc-192a982657a8.png)

In this case, **Expired** takes control and can continue with any flow.

**Comment by MaheshkumarSundaram** [comment]
Hi @ralphhanna,

As per standards (refer [camunda doc](https://camunda.com/bpmn/reference/#events-basic-concepts)), after an interrupting event occurs, the original pathway should not execute. Only the new path has to be followed.

In that case, in the original example,

When the first timer **Expired Timer** fires after specified time (1 minute) it will:

- End the Parent Activity **Receive**
- Which in turn will Cancel the **Notify** Timer
- Will invoke **Expired** ServiceTask
- Typically it should follow the path of the  **Expired** ServiceTask. In our case as there is nothing after **Expired** ServiceTask, the process should end.

Thanks!

**Comment by ralphhanna** [comment]
Thanks, that required one more fix, done part of new release

**Comment by MaheshkumarSundaram** [comment]
Hi @ralphhanna,

> Thanks, that required one more fix, done part of new release

Is this released? Thanks!

**Comment by ralphhanna** [comment]
Is Fixed in release 1.3.10 - ready now
